src/ompl/geometric/planners/rrt/src/STRRTstar.cpp

```diff
@@ -0,0 +1,1114 @@
+            continue;
+        }
+
+        // determine whether the old or new batch is sampled
+        sampleOldBatch_ =
+            (firstBatch || isTimeBounded_ || !sampleUniformForUnboundedTime_ || rng_.uniform01() <= oldBatchSampleProb);
+
+        ompl::base::State *goalState{nullptr};
+        if (sampleOldBatch_)
+        {
+            // sample until successful or time is up
+            if (goalMotions_.empty() && isTimeBounded_)
+                goalState = nextGoal(ptc, oldBatchTimeBoundFactor, newBatchTimeBoundFactor);
+            // sample for n tries, with n = batch size
+            else if (goalMotions_.empty() && !isTimeBounded_)
+            {
+                goalState = nextGoal(static_cast<int>(batchSize), oldBatchTimeBoundFactor, newBatchTimeBoundFactor);
+                // the goal region is most likely blocked for this time period -> increase upper time bound
+                if (goalState == nullptr)
+                {
+                    newBatchTimeBoundFactor *= timeBoundFactorIncrease_;
+                    oldBatchTimeBoundFactor = newBatchTimeBoundFactor;
+                    startTree = true;
+                    if (sampleUniformForUnboundedTime_)
+                        batchSize = std::ceil(2.0 * (timeBoundFactorIncrease_ - 1.0) *
+                                              static_cast<double>(tStart_->size() + tGoal_->size()));
+                    numBatchSamples = 0;
+                    OMPL_INFORM("%s: Increased time bound factor to %.2f", getName().c_str(), newBatchTimeBoundFactor);
+                    continue;
+                }
+            }
+            // sample for a single try
+            else if (forceGoalSample ||
+                     goalMotions_.size() < (tGoal_->size() - newBatchGoalSamples) / goalStateSampleRatio_)
+            {
+                goalState = nextGoal(1, oldBatchTimeBoundFactor, newBatchTimeBoundFactor);
+                forceGoalSample = false;
+            }
+        }
+        else
+        {
+            if (newBatchGoalMotions_.empty())
+            {
+                goalState = nextGoal(static_cast<int>(batchSize), oldBatchTimeBoundFactor, newBatchTimeBoundFactor);
+                // the goal region is most likely blocked for this time period -> increase upper time bound
+                if (goalState == nullptr)
+                {
+                    oldBatchTimeBoundFactor = newBatchTimeBoundFactor;
+                    newBatchTimeBoundFactor *= timeBoundFactorIncrease_;
+                    startTree = true;
+                    batchSize = std::ceil(2.0 * (timeBoundFactorIncrease_ - 1.0) *
+                                          static_cast<double>(tStart_->size() + tGoal_->size()));
+                    numBatchSamples = 0;
+                    OMPL_INFORM("%s: Increased time bound factor to %.2f", getName().c_str(), newBatchTimeBoundFactor);
+                    continue;
```

Comment:
You should avoid duplicate code. So just put it into a function and call it from here and the place before.